overlord/snapstate/handlers.go

```diff
@@ -157,55 +170,89 @@ func (m *SnapManager) doPrerequisites(t *state.Task, _ *tomb.Tomb) error {
 		return nil
 	}
 
-	// check prereqs
-	prereqName := defaultCoreSnapName
+	// we need to make sure we install all prereqs together in one
+	// operation
+	base := defaultCoreSnapName
 	if snapsup.Base != "" {
-		prereqName = snapsup.Base
+		base = snapsup.Base
 	}
-
-	var prereqState SnapState
-	err = Get(st, prereqName, &prereqState)
-	// we have the prereq already
-	if err == nil {
-		return nil
-	}
-	// if it is a real error, report
-	if err != state.ErrNoState {
+	if err := m.installPrereqs(t, base, snapsup.Prereq, snapsup.UserID); err != nil {
 		return err
 	}
 
-	// check that there is no task that installs the prereq already
-	prereqPending, err := changeInFlight(st, prereqName)
+	return nil
+}
+
+func (m *SnapManager) installOnePrereq(st *state.State, prereqName string, userID int) (*state.TaskSet, error) {
+	// installed already?
+	isInstalled, err := isInstalled(st, prereqName)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if prereqPending {
-		// if something else installs core already we need to
-		// wait for that to either finish successfully or fail
-		return &state.Retry{After: prerequisitesRetryTimeout}
+	if isInstalled {
+		return nil, nil
+	}
+	// in progress?
+	inFlight, err := linkSnapInFlight(st, prereqName)
+	if err != nil {
+		return nil, err
+	}
+	if inFlight {
+		return nil, nil
```

Comment:
if the prereq is the the base we need to wait also in this case, are we missing some test?